src/pb_stub.cc

```diff
@@ -837,6 +864,183 @@ Stub::GetOrCreateInstance()
+  }
+  log_message_cv_.notify_one();
+}
+
+void
+Stub::ServiceLogRequests()
+{
+  while (log_thread_) {
+    std::unique_lock<std::mutex> guard{log_message_mutex_};
+    while (log_request_buffer_.empty()) {
+      log_message_cv_.wait(guard);
+    }
+    // In case of exit, do not send message
+    if (log_thread_) {
+      std::unique_ptr<PbLog> log_request =
+          std::move(log_request_buffer_.front());
+      log_request_buffer_.pop();
+      SendLogMessage(log_request);
+    }
+  }
+}
+
+void
+Stub::SendLogMessage(std::unique_ptr<PbLog>& log_send_message)
+{
+  std::unique_ptr<PbLogShm> log_request_shm = PbLogShm::Create(
+      shm_pool_, log_send_message->Filename(), log_send_message->Line(),
+      log_send_message->Message(), log_send_message->Level());
+  LogSendMessage* send_message_payload = log_request_shm->LogMessage();
+  send_message_payload->waiting_on_stub = false;
+  std::unique_ptr<IPCMessage> log_request_msg =
+      IPCMessage::Create(shm_pool_, false /* inline_response */);
+  log_request_msg->Args() = log_request_shm->ShmHandle();
+  ScopedDefer _([send_message_payload] {
+    {
+      bi::scoped_lock<bi::interprocess_mutex> guard{
+          send_message_payload->log_mu};
+      send_message_payload->waiting_on_stub = false;
+      send_message_payload->log_cv.notify_all();
+    }
+  });
+
+  {
+    // Send a message to be caught by the log monitor thread in python_be.cc
+    bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->log_mu};
+    SendIPCLogMessage(log_request_msg);
+    while (!send_message_payload->waiting_on_stub) {
+      send_message_payload->log_cv.wait(guard);
+    }
+  }
+}
+
+bool
+Stub::LogServiceActive()
+{
+  return log_thread_;
+}
+
+// Bound function, called from the python client
+void
+Logger::Log(const std::string& message, LogLevel level)
+{
+  std::unique_ptr<Stub>& stub = Stub::GetOrCreateInstance();
+  py::object frame = py::module_::import("inspect").attr("currentframe");
+  py::object caller_frame = frame();
+  py::object info = py::module_::import("inspect").attr("getframeinfo");
+  py::object caller_info = info(caller_frame);
+  py::object filename_python = caller_info.attr("filename");
+  std::string filename = filename_python.cast<std::string>();
+  py::object lineno = caller_info.attr("lineno");
+  uint32_t line = lineno.cast<uint32_t>();
+
+  if (!stub->LogServiceActive()) {
+    Logger::GetOrCreateInstance().Log(filename, line, level, message);
+  } else {
+    std::unique_ptr<PbLog> log_msg(new PbLog(filename, line, message, level));
+    stub->EnqueueLogRequest(log_msg);
+  }
+}
+
+// Called internally (.e.g. LOG_ERROR << "Error"; )
+void
+Logger::Log(
+    const std::string& filename, uint32_t lineno, LogLevel level,
+    const std::string& message)
+{
+  std::unique_ptr<Stub>& stub = Stub::GetOrCreateInstance();
+  // If the log monitor service is not active yet, format
+  // and pass messages to cerr
+  if (!stub->LogServiceActive()) {
```

Comment:
I was wondering why LogService is not active. Is there anyway to make it active before we call the user's `Initialize` function.

Reply:
There are 2 critical periods where log service is not active, startup and shutdown. During startup, we can only start the logging threads after the stub launcher has allocated the log message queue. The reason I wait until the stub is initialized before activating the log thread is because activating it before assumes that the launch will be successful and that the backend will begin listening for messages from the stub. This is not a safe guarantee and will result in a deadlock if a message is pushed onto the queue but the stub fails to launch. This is the same reason I terminate the stub log thread before any stub termination messages are sent. The Python backend must terminate its log monitor prior to terminating the stub because the stub termination process will deallocate the log message queue. If we terminate the stub log thread last, we will also encounter a deadlock for the same reason. 

I figure the safest handling would be to have the condition variables do a wait_for and check their predicates on timeout. But I was worried this would degrade performance, especially if the user does not post any log messages.

Reply:
I see. Thanks for the detailed response!

Reply:
Just wanted to make sure that the user can use logging in `initialize` and `finalize` parts of their Python model.

Reply:
The user can access the pybind functions with no issues when logging during initialize and finalize. However, the messages will be logged using cerr as opposed to triton logging as the threads will not have started up yet/ will have already terminated.

Reply:
Resolving, this is now outdated. User can log messages to Triton within initialize and finalize as well now.